aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -35,43 +35,92 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
 #else // AT_MKLDNN_EBABLED
 
 #include <ATen/mkldnn/Runtime.h>
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
 
 using namespace mkldnn;
 
 namespace at { namespace native {
 
-constexpr int input_batch_size_dim = 0;  // also grad_input
-constexpr int input_channels_dim = 1;
-constexpr int output_batch_size_dim = 0;  // also grad_output
-constexpr int output_channels_dim = 1;
-constexpr int weight_output_channels_dim = 0;
-constexpr int weight_input_channels_dim = 1;
+Tensor _ideep_conv2d(
+    const Tensor& input, const Tensor& weight, const Tensor& bias,
+    at::IntArrayRef padding, at::IntArrayRef stride, at::IntArrayRef dilation, int64_t groups) {
 
-// Often written as 2 + max_dim (extra dims for batch size and channels)
-constexpr int max_dim = 3;
+  auto stride_vec = expand_param_if_needed(stride, 2);
+  auto padding_vec = expand_param_if_needed(padding, 2);
+  auto dilation_vec = expand_param_if_needed(dilation, 2);
 
-static std::vector<int64_t> conv_output_size(
-    IntArrayRef input_size, IntArrayRef weight_size,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto dim = input_size.size();
-  std::vector<int64_t> output_size(dim);
-  output_size[0] = input_size[input_batch_size_dim];
-  output_size[1] = weight_size[weight_output_channels_dim];
-  for (size_t d = 2; d < dim; ++d) {
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+
+  std::vector<int64_t> kernel_size(input.dim());
+  if (groups > 1 && (weight.dim() == input.dim() + 1)) {
+    kernel_size[0] = weight.size(0) * weight.size(1);
```

Comment:
what's going on here?  Why is the `conv_output_size` used here like the other ones but with different parameter names?  What's going on with the kernel_size here?

Reply:
This is related `mkldnn_reorder_conv2d_weight`, which converts the conv weights from 4d to 5d.